README.md
# node-pub-sub

<a href="https://nodei.co/npm/node-pub-sub/"><img src="https://nodei.co/npm/node-pub-sub.png"></a>

[![Build Status](https://travis-ci.org/joaquimserafim/node-pub-sub.png?branch=master)](https://travis-ci.org/joaquimserafim/node-pub-sub)



Implements Publish/Subscribe messaging paradigm in a very simple way.


**V1.3**

  
### Some Wiki

In software architecture, publish–subscribe is a messaging pattern where senders of messages, called publishers, do not program the messages to be sent directly to specific receivers, called subscribers. Instead, published messages are characterized into classes, without knowledge of what, if any, subscribers there may be. Similarly, subscribers express interest in one or more classes, and only receive messages that are of interest, without knowledge of what, if any, publishers there are.

Pub/sub is a sibling of the message queue paradigm, and is typically one part of a larger message-oriented middleware system.

Most messaging systems support both the pub/sub and message queue models in their API, e.g. Java Message Service (JMS).



### API

All communication/messages it's done in JSON.

#### Publish
    
    require('node-pub-sub').Publish
      
    Publish([port], [host], [options])
      
     port: default to 5000
     host: default to localhost
     options: {timeout: 120000, encoding: 'ascii'}
          timeout: put a timeout like 1 hour or 6 hours because if spend much 
              time without publish messages the server will close the 
              connection, default to 3600000
          
      
    events:
          ('ready', callback)
          ('error', callback (err))
          ('end', callback)
    
    
    // publish messages
    pub.publish('some channel', 'some message', [callback])
    
    // end
    pub.end()

#### Subscribe

    require('node-pub-sub').Publish
    
    Publish(port, [host], [options])
    
      host: default to localhost
      port: default to 5000
      options: {timeout: 120000, encoding: 'ascii'}
          timeout: put a timeout like 1 hour or 6 hours because if spend much
              time without publish messages the server will close the connection,
              default to 3600000
          
    
    // subscribe to a channel      
    sub.start('some channel', callback ('error/warn message', 'message - json object'))
    
    // unsubscribe, for now unsubscribe to all channels
    sub.end()


### Examples

    **Publish**
    
    var Publish = require('node-pub-sub').Publish;
    
    var pub = new Publish(5000, {timeout: 120000});
    
    pub.on('ready', function () { console.log('publish ready'); });
    pub.on('error', function (err) { console.log(err); });
    pub.on('end', function () { console.log('publish ended'); });
    

    pub.publish('some channel', 'some message', function () {
        // some acknowledgement
    });
    
    // close Publish    
    pub.end();
    
    
    Message format:
    {"n":990,"tmx":1392154487968,"msg":"Hello World"}
    
    create by Publish
    n   => message counter 
    tmx => message timestamp
    
    
    **Subscribe**
    
    var Subscribe = require('node-pub-sub').Subscribe;
    
    var sub = new Subscribe(5000, {timeout: 120000});
    
    sub.start(channel, function (err, message) {
        if (err) throw err;
        
        // JSON object 
        console.log(message);
    });

    // unsubscribe when you want
    setTimeout(function () {
        sub.end();
    }, 60000);
    
    




ChangeLog.md
2013.12.30, Version 1.0.0

* first release


2014.01.12, Version 1.0.1

* Removed support to Node.js 0.8
* Update split-json module


2014.01.29, Version 1.1.1

* Add "host" parameter to Publish enabling to define the ip that Publish will listen

2014.02.02, Version 1.2.1

* Remove 'encoding' option from Publish and Subscriber
* Update dependencies


2014.02.02, Version 1.2.2

* Use readable-stream


2014.02.03, Version 1.2.3

* Pub/Sub - Fix in module node-netcat when send a 'null' value.


2014.02.11, Version 1.3.3

* Add module 'tiny-eventemitter' for EventEmitter
* Update dependencies
* Remove event 'close' in Subscribe

test/index.js
var test = require('tape');
var Guid = require('guid');


function JSONValid (data) {
  try {
    return JSON.parse(data);
  } catch (err) {
    return null;
  }
}

var PubSub = require('../');

var Publish = PubSub.Publish;
var Subscribe = PubSub.Subscribe;

var num_t = 2002;

test('pub-sub', function (t) {
  t.plan(num_t);

  var pub = new Publish(7777, '127.0.0.1', {timeout:120000, encoding: 'ascii'});
  var sub = new Subscribe(7777, '127.0.0.1', {timeout: 120000, encoding: 'ascii'})
  
  pub.on('ready', function () { t.pass('publish, ready'); });
  pub.on('error', function (err) { t.error(err !== null, err); });
  pub.on('end', function () { t.pass('publish, ended'); });


  var channel = 'xpto';

  // create some messages
  var messages = [];

  for (var i = 0; i < 1000; i++)
    messages.push('Hello World ' + Guid.raw());

  var control_exit = messages.length;

  // subscribing
  sub.start(channel, function (err, obj) {
    if (err) return t.ok(err, err);
    //t.ok(_.isObject(obj), JSONhandler.stringify(obj));
    t.pass(obj);
    if (--control_exit === 0) {
      sub.end();
      setTimeout(function () { pub.end(); }, 500);
    }
  });


  // publishing
  setTimeout(function () {
    messages.forEach(function (msg) {
      pub.publish(channel, msg, function () {
        t.ok(msg, msg);
      });
    });
  }, 2000);
});

lib/sub.js
var _ = require('underscore');
var Netcat = require('node-netcat');
var Readable = require('stream').Readable || require('readable-stream').Readable;
var split = require('split-json');


module.exports = Subscribe;

function Subscribe (port, host, options) {
  if (!(this instanceof Subscribe)) return new Subscribe(port, host, options);
  Subscribe.init.call(this, port, host, options);
}

Subscribe.init = function (port, host, options) {
  // check args
  if (_.isObject(host)) {
    options = host;
    host = 'localhost';
  }

  this._port = port; 
  this._host = host || 'localhost';
  this._timeout = (options && options.timeout) || 3600000;

  this._client = Netcat.client(this._port, this._host, {timeout: this._timeout});
};

Subscribe.prototype.end = function () {
  return this._client.send(null, true);
};

Subscribe.prototype.start = function (channel, cb) {
  var self = this;

  var json = JSON.stringify({channel: channel});

  // events callbacks

  function open () { 
    setTimeout(function cb_send_channel () {
      self._client.send(json); 
    }, 500);
  }

  function error (err) { cb(err); }
  
  function data (data) {
    // create a readable stream
    var rs = new Readable();
    rs.push(data);
    rs.push(null);

    rs.pipe(split(/\n/)).on('data', function cb_data (data) { cb(null, data); });
  }

  // events

  this._client.on('open', open);
  this._client.on('data', data);
  this._client.on('error', error);

  this._client.start();
};

lib/pub.js
var inherits = require('util').inherits;
var EventEmitter = require('tiny-eventemitter');
var _ = require('underscore');
var Netcat = require('node-netcat');
var split = require('split-json');



module.exports = Publish;

function Publish (port, host, options) {
  if (!(this instanceof Publish)) return new Publish(port, host, options);
  EventEmitter.call(this);
  Publish.init.call(this, port, host, options);
}

inherits(Publish, EventEmitter);


Publish.init = function (port, host, options) {
  var self = this;

  // check the options

  if (_.isObject(host)) {
    options = host;
    host = 'localhost';
  }

  this._port = port || 5000;
  this._host = host || 'localhost';
  this._timeout = (options && options.timeout) || 3600000;

  // internal message numerator
  this._count = 0;

  this._channels = {};
  this._ready = false;// server is ready

  this._server = new Netcat.server(this._port, this._host, {timeout: this._timeout});

  // events callbacks

  function subscribe (client, data) {
    var json = split.JSONValid(data.toString());

    // not valid, exit
    if (!json) return self._server.send(client, '' , true);

    // client is subscribing a channel does not exist yet
    if (_.isEmpty(self._channels[json.channel])) {
      self._channels[json.channel] = {clients: [client]};
      return;// exit
    }

    // check client is already in subscribed in channel
    var exists = _.contains(self._channels[json.channel].clients, client);
    // no, then save client in channel list
    if (!exists) self._channels[json.channel].clients.push(client);
  }

  function unSubscribe (client) {
    function rm (channel) { channel.clients.splice(client, 1); }
    _.each(self._channels, rm);
  }

  function ready () { self.emit('ready'); }

  function error (err) { self.emit('error', err); }

  function end () { self.emit('end'); }

  // server events
  this._server.on('ready', ready);
  this._server.on('data', subscribe);
  this._server.on('client_off', unSubscribe);
  this._server.on('error', error);
  this._server.on('close', end);

  // emit to publish the messages
  this.on('new_msg', self._process);

  this._server.listen();
};

Publish.prototype.clients = function () {
  return this._server.getClients();
};

Publish.prototype.end = function () {
  // lets close conn with clients
  _.each(this._channels, function channels (channel) { 
    _.each(channel.clients, function clients (client) {
      this._server.send(client, null, true);
    });
  });
  // now close server
  this._server.close();
};

Publish.prototype._process = function (channel, msg, cb) {
  var self = this;

  function processMsg (client) { self._server.send(client, msg, false, cb); }
  // get the clients for the channel and send message
  _.each(this._channels[channel].clients, processMsg);
};

Publish.prototype.publish = function (channel, msg, cb) {
  // check callback
  cb = cb || function () {};
  // new channel init object
  if (_.isEmpty(this._channels[channel])) this._channels[channel] = {clients: []};

  // inc numerator
  this._count++;  

  var data = JSON.stringify({n: this._count, tmx: new Date().getTime(), msg: msg}) + '\n';
  // emit the new message to be published
  this.emit('new_msg', channel, data, cb);
};